File: src/utils/logger.rb
class Logger

  def initialize(log_level=:error)
    @log_level = log_level
  end

  def error(message)
    if can_print(:error)
      puts message
    end
  end

  def info(message)
    if can_print(:info)
      puts message
    end
  end

  def debug(message)
    if can_print(:debug)
      puts message
    end
  end

  def trace(message)
    if can_print(:trace)
      puts message
    end
  end

  private
  def can_print(method)
    if method == :error
      true
    elsif method == :info
      return @log_level != :error
    elsif method == :debug
      return (@log_level == :debug or @log_level == :trace)
    else
      @log_level == :trace
    end
  end
end

File: sender/utils/logging/Logger.ts

import {LogLevel} from "./LogLevel";

export abstract class Logger {
    private readonly level: LogLevel;

    protected constructor(logLevel: LogLevel) {
        this.level = logLevel;
    }

    abstract log(message: any): void;

    error(message: any, er?: Error): void {
        if (this.canLog(LogLevel.Error)) {
            this.log(message);
            if (er != null) {
                this.log(er);
            }
        }
    }

    info(message: any): void {
        if (this.canLog(LogLevel.Info)) {
            this.log(message);
        }
    }

    debug(message: any): void {
        if (this.canLog(LogLevel.Debug)) {
            this.log(message);
        }
    }

    trace(message: any): void {
        if (this.canLog(LogLevel.Trace)) {
            this.log(message);
        }
    }

    private canLog(methodLevel: LogLevel) {
        return methodLevel >= this.level;
    }
}

export { LogLevel }